Helper/Discount.php

```diff
@@ -0,0 +1,320 @@
+     */
+    private $resource;
+
+    /**
+     * @var ThirdPartyModuleFactory
+     */
+    protected $amastyAccountFactory;
+
+    /**
+     * @var ThirdPartyModuleFactory
+     */
+    private $amastyGiftCardManagement;
+
+    /**
+     * @var ThirdPartyModuleFactory
+     */
+    protected $amastyQuoteFactory;
+
+    /**
+     * @var ThirdPartyModuleFactory
+     */
+    protected $amastyQuoteResource;
+
+    /**
+     * @var ThirdPartyModuleFactory
+     */
+    protected $amastyQuoteRepository;
+
+    /**
+     * @var CartRepositoryInterface
+     */
+    protected $quoteRepository;
+
+    /**
+     * @var ConfigHelper
+     */
+    private $configHelper;
+
+    /**
+     * Discount constructor.
+     *
+     * @param Context $context
+     * @param ResourceConnection $resource
+     * @param ThirdPartyModuleFactory $amastyAccountFactory
+     * @param ThirdPartyModuleFactory $amastyGiftCardManagement
+     * @param ThirdPartyModuleFactory $amastyQuoteFactory
+     * @param ThirdPartyModuleFactory $amastyQuoteResource
+     * @param ThirdPartyModuleFactory $amastyQuoteRepository
+     * @param CartRepositoryInterface $quoteRepository
+     * @param ConfigHelper $configHelper
+     *
+     * @codeCoverageIgnore
+     */
+    public function __construct(
+        Context $context,
+        ResourceConnection $resource,
+        ThirdPartyModuleFactory $amastyAccountFactory,
+        ThirdPartyModuleFactory $amastyGiftCardManagement,
+        ThirdPartyModuleFactory $amastyQuoteFactory,
+        ThirdPartyModuleFactory $amastyQuoteResource,
+        ThirdPartyModuleFactory $amastyQuoteRepository,
+        CartRepositoryInterface $quoteRepository,
+        ConfigHelper $configHelper
+    ) {
+        parent::__construct($context);
+        $this->resource = $resource;
+        $this->amastyAccountFactory = $amastyAccountFactory;
+        $this->amastyGiftCardManagement = $amastyGiftCardManagement;
+        $this->amastyQuoteFactory = $amastyQuoteFactory;
+        $this->amastyQuoteResource = $amastyQuoteResource;
+        $this->amastyQuoteRepository = $amastyQuoteRepository;
+        $this->quoteRepository = $quoteRepository;
+        $this->configHelper = $configHelper;
+    }
+
+    /**
+     * Collect and update quote totals.
+     * @param Quote $quote
+     */
+    protected function updateTotals(Quote $quote)
+    {
+        $quote->getShippingAddress()->setCollectShippingRates(true);
+        $quote->setTotalsCollectedFlag(false);
+        $quote->collectTotals();
+        $quote->setDataChanges(true);
+        $this->quoteRepository->save($quote);
+    }
+
+    /**
+     * Load Amasty Gift Card account object.
+     * @param string $code Gift Card coupon code
+     * @return \Amasty\GiftCard\Model\Account|null
+     */
+    public function loadAmastyGiftCard($code)
+    {
+        try {
+            $accountModel = $this->amastyAccountFactory->getInstance()
+                ->create()
+                ->loadByCode($code);
```

Comment:
Does Amasty have something like Repository instead of getInstance()->createe()->loadByCode()?

Reply:
@YevhenShBolt It uses Factories and Resource Models. getInstance() is our ThirdPartyModuleFactory call for (trying) instantiating third party module objects that might not be present (installed). I think it's your (great) solution. The rest is more or less copied from the module, apply controller in this particular case. That is how it's done in the Amasty module.

Reply:
Thank you, now i see.